spikeinterface/sortingcomponents/matching/wobble.py

```diff
@@ -0,0 +1,948 @@
+        spike_train[:, 1] //= params.jitter_factor
+
+        # TODO : Benchmark spike amplitudes
+        # Find spike amplitudes / channels
+        amplitudes, channel_inds = [], []
+        for i, spike_idx in enumerate(spike_train[:, 0]):
+            best_ch = np.argmax(np.abs(traces[spike_idx, :]))
+            amp = np.abs(traces[spike_idx, best_ch])
+            amplitudes.append(amp)
+            channel_inds.append(best_ch)
+
+        # assign result to spikes array
+        spikes = np.zeros(spike_train.shape[0], dtype=cls.spike_dtype)
+        spikes['sample_ind'] = spike_train[:, 0]
+        spikes['cluster_ind'] = spike_train[:, 1]
+        spikes['channel_ind'] = channel_inds
+        spikes['amplitude'] = amplitudes
+
+        return spikes
+
+
+    @classmethod
+    def run_array(cls, traces, template_meta, params, sparsity, objective):
+        """Performs template matching using the WobbleMatch algorithm.
+
+        Spike times are indexed at the beginning of the template rather than the trough of the spike.
+        Spike ids refer to the jittered ids rather than the unit ids.
+
+        Parameters
+        ----------
+        traces : ndarray (chunk_len + 2*margin, n_chan)
+            Voltage traces for a chunk of the recording.
+        template_meta : TemplateMetadata
+            Dataclass object for aggregating template metadata together.
+        params : WobbleParameters
+            Dataclass object for aggregating the parameters together.
+        sparsity : Sparsity
+            Dataclass object for aggregating channel sparsity variables together.
+        objective : Objective
+            Dataclass object for aggregating variables related to the objective together.
+
+        Returns
+        -------
+        spike_train : ndarray (n_spikes, 2)
+            Spike train from template matching.
+        scalings : ndarray (n_spikes,)
+            Amplitude scaling used for each spike.
+        distance_metric : ndarray (n_spikes)
+            A metric that describes how good of a 'fit' each spike is to its corresponding template
+        """
+        # Compute objective
+        compute_objective(traces, objective, params.conv_approx_rank, template_meta.n_templates)
+
+        # Compute spike train
+        spike_trains, scalings, distance_metrics = [], [], []
+        for i in range(params.max_iter):
+            # find peaks
+            spike_train, scaling, distance_metric = cls.find_peaks(objective, params, template_meta)
+            if len(spike_train) == 0:
+                break
+
+            # update spike_train, scaling, distance metrics with new values
+            spike_trains.extend(list(spike_train))
+            scalings.extend(list(scaling))
+            distance_metrics.extend(list(distance_metric))
+
+            # subtract newly detected spike train from traces (via the objective)
+            cls.subtract_spike_train(spike_train, scaling, objective, params, template_meta, sparsity)
```

Comment:
Does this line modifies the variable traces? 

I am confused as to how is this modified across the loop. Is `find_peaks` or `substract_spike_trains` implicitly modifying it? If so, is traces an attribute? 

I kind of feel that this functions should return something.

Reply:
The traces are not modified during the course of this algorithm.  Instead, it subtracts spikes from the objective itself (which is really just a convolution of the traces with the templates).  

The templates are also not altered during the course of template matching in this alg (which I think is for the best tbh).

So, the only 2 things that get updated every iteration are the objective (as well as the normalized version) and the spike train.

I hope that clears things up, lmk if you have more questions.